### PR TITLE
CRDCDH-639

### PR DESCRIPTION
**Pull request by knockknockyoo**
1. test codes including test cases
2. required properties validator

### CODE REVIEW

File: src/common/constants.py [hunk 1]
```diff
@@ -89,6 +89,8 @@
 # sqs message
 SCOPE = "scope"
 
-
-
+# validation status
```
Comment (vshand11) [inline src/common/constants.py:92]: These constants have been defined.
FILE_STATUS = "status"
STATUS_ERROR = "Error"
STATUS_WARNING = "Warning"
STATUS_PASSED = "Passed"
STATUS_NEW = "New"

File: src/metadata_validator.py [hunk 2]
```diff
@@ -149,8 +149,41 @@ def validateNode(self, dataRecord, model):
         
         return STATUS_PASSED, None, None
     
-    def validate_required_props(self, dataRecord, model):
-        return None
+    def validate_required_props(self, data_record, node_definition):
+        result = {"result": ERROR, "errors": [], "warnings": []}
+        # check the correct format from the node_definition
+        if "model" not in node_definition.keys() or "nodes" not in node_definition["model"].keys():
+            result["errors"].append(f"node definition is not correctly formatted.")
```
Comment (vshand11) [inline src/metadata_validator.py:156]: error should type of ErrorMessage,   warning use the same type.   Don't use plain string as error or warning
{ title: String description: String }

File: src/metadata_validator.py [hunk 3]
```diff
@@ -149,8 +149,41 @@ def validateNode(self, dataRecord, model):
...
+        anode = nodes[node_type]
+        if node_type in nodes.keys():
+            for key, value in data_record["rawData"].items():
+                anode_keys = anode.keys()
+                if "properties" not in anode_keys:
+                    result["errors"].append(f"data record is not correctly formatted.")
+                    break
+
+                key_not_exist = key not in anode["properties"]
+                value_invalid = not value.strip()
```
Comment (vshand11) [inline src/metadata_validator.py:180]: do yo need to check if the prop is required?

**Review by vshand11 — CHANGES_REQUESTED** [review]
Hi Young,  Please take a look on comments, some may need to change.

File: src/metadata_validator.py (hunk 2 above)
Comment (knockknockyoo) [reply src/metadata_validator.py]: updated it to error object

File: src/metadata_validator.py [hunk 4]
```diff
@@ -149,8 +149,41 @@ def validateNode(self, dataRecord, model):
...
+        # validation start
+        nodes = node_definition["model"]["nodes"]
+        node_type = data_record["nodeType"]
+        # extract a node from the data record
+        if node_type not in nodes.keys():
+            result["errors"].append(f"Required node '{node_type}' does not exist.")
+            return result
+
+        anode = nodes[node_type]
+        if node_type in nodes.keys():
```
[the remark this answers is not in the document]
Comment (knockknockyoo) [reply src/metadata_validator.py]: The primary goal of this PR is to check whether the necessary properties in the data_record are present and not empty. This is required.

File: src/metadata_validator.py (hunk 3 above)
Comment (knockknockyoo) [reply src/metadata_validator.py]: The primary goal of this PR is to check whether the necessary properties in the data_record are present and not empty. This is required.

File: src/metadata_validator.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (knockknockyoo) [reply src/metadata_validator.py]: This is deleted

**Comment by vshand11** [comment]
Hi Young,  get your new codes and find some variables are not defined from line 201 to 209 in metadata_validator.

**Comment by knockknockyoo** [comment]
Peter, I've pushed the commit for the changes.

**Review by vshand11 — APPROVED** [review]
New version codes passed review and tests.

**Pull request merged**